### PR TITLE
[bazel] Add `dual_cc_library` for simpler mockable targets.

### PR DESCRIPTION
**Pull request by mcy**
This PR gives us a new way to write "mockable" libraries that has only one target, rather than several, per logical library. This will simplify the declaration of dependencies for both unit tests and device binaries, without the current mess of mock/non-mock libraries and needing to include the sources of a library in multiple targets.

This PR rewrites the MMIO and CSR libraries (at the very bottom of the dependency graph) to use `dual_cc_library`, as an example of how they could be used. Unfortunately, due to how the code is structured on the Meson side, it's not possible to move files around quite yet to reflect the new structure, and will need to wait until we delete Meson.

Note that this PR actually deletes *several* targets that are now redundant.

### CODE REVIEW

File: rules/BUILD [hunk 1]
```diff
@@ -2,4 +2,11 @@
...
 # SPDX-License-Identifier: Apache-2.0
 
+load("//rules:opentitan.bzl", "OPENTITAN_PLATFORM")
+
 package(default_visibility = ["//visibility:public"])
+
+config_setting(
+    name = "opentitan_platform",
+    values = {"platforms": OPENTITAN_PLATFORM},
+)
```
Comment (timothytrippel) [inline rules/BUILD:12]: Is there any reason this cannot live in `rules/opentitan.bzl`? Seems more appropriate to be defined there?

File: meson.build [hunk 2]
```diff
@@ -106,6 +106,8 @@ c_cpp_args = [
   # We use absolute include paths as much as possible.
   '-I' + meson.project_source_root(),
   '-I' + meson.project_build_root(),
+  # Allow detection of Meson for migration purposes.
```
Comment (timothytrippel) [inline meson.build:109]: nit: a "FIXME" like you had with the bazel vendoring changes might make this easier to grep for when we migrate to bazel fully

**Review by timothytrippel — APPROVED** [review]
Overall, I like the idea! Should save some  LOC when writing on-device libs that need to be tested on-host.

File: rules/BUILD (hunk 1 above)
Comment (mcy) [reply rules/BUILD:12]: It can't. You can't define targets in .bzl files.

File: meson.build (hunk 2 above)
Comment (mcy) [reply meson.build:109]: I mean. It'll get deleted when we do `rm -rf **/meson.build`. So I don't think it's strictly necessary.

File: rules/cross_platform.bzl [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Rules for defining cross-platform libraries."""
+
+def dual_inputs(shared = [], device = [], host = []):
+    """Constructs a cross-platform input for use with dual_cc_library."""
+    return struct(shared = shared, device = device, host = host)
+
+def dual_cc_device_library(label):
```
[the remark this answers is not in the document]
Comment (mcy) [reply rules/cross_platform.bzl:11]: How about `dual_cc_device_library_of()`?

**Comment by mcy** [comment]
We chatted about this offline, but I did want to address this one:

> The independent MOCK_* defines suggest an intent to control mocks individually, and it seems like a good time to check we don't need that.

These predate the proliferation of other mocks, and exist mostly to work around bad life choices around Meson (I can attest to this because I wrote them :P). The plan is to replace all of them with Chris's platform macros, but that may need to wait until Meson is gone.

**Review by drewmacrae — APPROVED** [review]
LGTM, I'm assuming this will be aligned with the HOST/Target macros more than the RV32 one that imagines we might have more than 2 targets. I think it's reasonable to imagine people will think about and write target specific code for about one at a time and that this code can stay out of the management of which flavor of silicon target the code is being built for.

**Review by cfrantz — APPROVED** [review]

**Comment by mcy** [comment]
Alphan: naming changes can be dealt with in parallel. I like off/on target, so I'll move to those in the future.

**Pull request merged**